src/circuit/gadget/poseidon/pow5.rs

```diff
@@ -87,123 +87,96 @@ impl<F: FieldExt> Pow5T3Chip<F> {
+                        let state_cur = meta.query_advice(state[idx], Rotation::cur());
+                        let rc_a = meta.query_fixed(rc_a[idx], Rotation::cur());
+                        acc + pow_5(state_cur + rc_a) * m_reg[next_idx][idx]
+                    });
+                    s_full.clone() * expr
+                })
+                .collect::<Vec<_>>()
         });
 
         meta.create_gate("partial rounds", |meta| {
             let cur_0 = meta.query_advice(state[0], Rotation::cur());
-            let cur_1 = meta.query_advice(state[1], Rotation::cur());
-            let cur_2 = meta.query_advice(state[2], Rotation::cur());
             let mid_0 = meta.query_advice(partial_sbox, Rotation::cur());
-            let next_0 = meta.query_advice(state[0], Rotation::next());
-            let next_1 = meta.query_advice(state[1], Rotation::next());
-            let next_2 = meta.query_advice(state[2], Rotation::next());
 
             let rc_a0 = meta.query_fixed(rc_a[0], Rotation::cur());
-            let rc_a1 = meta.query_fixed(rc_a[1], Rotation::cur());
-            let rc_a2 = meta.query_fixed(rc_a[2], Rotation::cur());
             let rc_b0 = meta.query_fixed(rc_b[0], Rotation::cur());
-            let rc_b1 = meta.query_fixed(rc_b[1], Rotation::cur());
-            let rc_b2 = meta.query_fixed(rc_b[2], Rotation::cur());
 
             let s_partial = meta.query_selector(s_partial);
 
-            let partial_round_linear = |idx: usize, rc_b: Expression<F>| {
-                s_partial.clone()
-                    * (mid_0.clone() * m_reg[idx][0]
-                        + (cur_1.clone() + rc_a1.clone()) * m_reg[idx][1]
-                        + (cur_2.clone() + rc_a2.clone()) * m_reg[idx][2]
-                        + rc_b
-                        - (next_0.clone() * m_inv[idx][0]
-                            + next_1.clone() * m_inv[idx][1]
-                            + next_2.clone() * m_inv[idx][2]))
+            use halo2::plonk::VirtualCells;
+            let mid = |idx: usize, meta: &mut VirtualCells<F>| {
+                let mid = mid_0.clone() * m_reg[idx][0];
+                (1..WIDTH).fold(mid, |acc, cur_idx| {
+                    let cur = meta.query_advice(state[cur_idx], Rotation::cur());
+                    let rc_a = meta.query_fixed(rc_a[cur_idx], Rotation::cur());
+                    acc + (cur + rc_a) * m_reg[idx][cur_idx]
+                })
             };
 
-            vec![
-                (
-                    "state[0] round a",
+            let next = |idx: usize, meta: &mut VirtualCells<F>| {
+                let next_0 = meta.query_advice(state[0], Rotation::next());
+                let next_0 = next_0 * m_inv[idx][0];
+                (1..WIDTH).fold(next_0, |acc, next_idx| {
+                    let next = meta.query_advice(state[next_idx], Rotation::next());
+                    acc + next * m_inv[idx][next_idx]
+                })
```

Comment:
As this covers the end of the second round in the row, we don't need to special-case the first state word:
```suggestion
                (0..WIDTH)
                    .map(|next_idx| {
                        let next = meta.query_advice(state[next_idx], Rotation::next());
                        next * m_inv[idx][next_idx]
                    })
                    .reduce(|acc, next| acc + next)
                    .expect("WIDTH > 0")
```